File: solutions/go/parsing-log-files/1/parsing_log_files.go
package parsinglogfiles

import (
    "regexp"
    "fmt"
)

func IsValidLine(text string) bool {
	re := regexp.MustCompile(`^\[([A-Z]{3})\]`)
    s := re.FindStringSubmatch(text)

    return len(s) == 2 && (
        s[1] == "TRC" ||
        s[1] == "DBG" ||
        s[1] == "INF" ||
        s[1] == "WRN" ||
        s[1] == "ERR" ||
        s[1] == "FTL")
}

func SplitLogLine(text string) []string {
	re := regexp.MustCompile(`<[~\*=\-]*>`)
    return re.Split(text, -1)
}

func CountQuotedPasswords(lines []string) int {
    re := regexp.MustCompile(`(?i)".*password.*"`)
	count := 0
    for _, line := range lines {
        if re.MatchString(line) {
            count += 1
        }
    }

    return count
}

func RemoveEndOfLineText(text string) string {
    re := regexp.MustCompile(`end-of-line\d+`)
    return re.ReplaceAllString(text, "")
}

func TagWithUserName(lines []string) []string {
	re := regexp.MustCompile(`User\s+(\S+)`)
    for i, line := range lines {
        s := re.FindStringSubmatch(line)
        if len(s) == 2 {
            lines[i] = fmt.Sprintf("[USR] %s %s", s[1], line)
        }
    }

    return lines
}

File: solutions/csharp/parsing-log-files/1/ParsingLogFiles.cs
using System;
using System.Text.RegularExpressions;
using System.Collections.Generic;

public class LogParser
{
    public bool IsValidLine(string text)
    {
        return Regex.IsMatch(text, @"^\[(TRC|DBG|INF|WRN|ERR|FTL)\]");
    }

    public string[] SplitLogLine(string text)
    {
        return Regex.Split(text, @"<[\^\*\=\-]*>");
    }

    public int CountQuotedPasswords(string lines)
    {
        return Regex.Count(lines, @"\"".*password.*\""", RegexOptions.IgnoreCase | RegexOptions.Multiline);
    }

    public string RemoveEndOfLineText(string line)
    {
        return Regex.Replace(line, @"end-of-line\d*", "");
    }

    public string[] ListLinesWithPasswords(string[] lines)
    {
        var result = new List<string>();
        foreach (var line in lines) {
            var m = Regex.Match(line, @"(password\w+)", RegexOptions.IgnoreCase);
            if (m.Success) {
                result.Add($"{m.Groups[1]}: {line}");
            } else {
                result.Add($"--------: {line}");
            }
        }

        return result.ToArray();
    }
}
